[PATCH v7] api: cls: add PMR priority to break PMR matching ties

JannePeltonen opened the pull request:
Add configurable priority to PMRs so that when multiple PMRs match a packet a PMR with the highest priority is selected. If multiple PMRs of the same priority match a packet, leave it implementation specific which PMR gets selected. Add a capability for the maximum priority value.

kkirankk commented on include/odp/api/spec/classification.h:
```diff
@@ -342,6 +342,18 @@ typedef struct odp_pmr_create_opt_t {
+	 *  Larger priority value corresponds to higher priority. The value
+	 *  must not be larger than odp_cls_capability_t::max_pmr_priority.
+	 *  The default value is 0.
+	 */
```
So, in case of chained COS, which rule should take priority? Is it PMR with high priority or PMR with longest chain?

JannePeltonen replied:
I am not sure if I understand the question. What do you mean with "PMR with the longest chain"?

CoSes from a tree that is rooted at a packet input. Coses are linked together through PMRs. Each CoS has zero or more PMRs that match packets that have reached the CoS. A matching PMR forwards a packet to its destination CoS. When a packet does not match any PMRs of a Cos, the packet has reached the final CoS after traversing a chain of CoSes. The action of the final CoS determines what happens to the packet.

When a packet matches multiple PMRs of a CoS, one of the PMRs has to be selected as the one that forwards the packet to the next CoS. Previously the selection was implementation specific. This PR adds priority to PMRs so that the highest priority matching PMR is the one that must be selected (and if there are multiple, the selection between them is still implementation specific).

The number of terms in a PMR does not make a difference in PMR selection (except possibly when the selection is implementation specific).

kkirankk replied:
Acked-by: Kiran Kumar Kokkilagadda <kirankumark@marvell.com>

JannePeltonen commented:
v2:
- very slightly tweaked API text in one place (no semantic change)
- added minimal implementation with max priority zero (i.e. no PMR prioritization support)
- added tests that were tested with a non-minimal PMR priority implementation

psavol reviewed (approved):
Reviewed API commit.

JannePeltonen commented:
v6: added review tag, rebased
